### PR TITLE
Migrate masked_scatter_ CUDA to ATen

### LINKED ISSUE #49542: Migrate `masked_scatter` from TH to ATen (CUDA)
Porting TH operators is essential for code simplicity and performance reasons.

Porting guides and Q&A are available in umbrella issue: #24507

Feel free to add @VitalyFedyunin as a reviewer to get a prioritized review.

cc @ezyang @gchanan @zou3519 @bdhirsh

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes #49542

### CODE REVIEW

**Comment by kshitij12345** [comment]
**Benchmark**

**Before PR**
```
Using pytorch 1.8.0a0+1373d82
[ masked_scatter_ mdtype torch.uint8 input dtypetorch.int32 ]
                          |  torch.masked_scatter_ CUDA
1 threads: --------------------------------------------
      (32,)               |             81.8           
      (32, 32)            |             87.6           
      (2, 16, 32)         |             85.2           
      (2, 16, 32, 32)     |             93.5           
      (4, 2, 16, 32, 32)  |             94.8           

Times are in microseconds (us).

[ masked_scatter_ mdtype torch.uint8 input dtypetorch.float32 ]
                          |  torch.masked_scatter_ CUDA
1 threads: --------------------------------------------
      (32,)               |             81.0           
      (32, 32)            |             83.0           
      (2, 16, 32)         |             90.4           
      (2, 16, 32, 32)     |             93.9           
      (4, 2, 16, 32, 32)  |             94.3           

Times are in microseconds (us).

[ masked_scatter_ mdtype torch.uint8 input dtypetorch.float64 ]
                          |  torch.masked_scatter_ CUDA
1 threads: --------------------------------------------
      (32,)               |             87.9           
      (32, 32)            |             89.5           
      (2, 16, 32)         |             89.8           
      (2, 16, 32, 32)     |             94.1           
      (4, 2, 16, 32, 32)  |             93.8           

Times are in microseconds (us).

[ masked_scatter_ mdtype torch.bool input dtypetorch.int32 ]
                          |  torch.masked_scatter_ CUDA
1 threads: --------------------------------------------
      (32,)               |             60.0           
      (32, 32)            |             62.3           
      (2, 16, 32)         |             62.7           
      (2, 16, 32, 32)     |             68.7           
      (4, 2, 16, 32, 32)  |             68.8           

Times are in microseconds (us).

[ masked_scatter_ mdtype torch.bool input dtypetorch.float32 ]
                          |  torch.masked_scatter_ CUDA
1 threads: --------------------------------------------
      (32,)               |             55.9           
      (32, 32)            |             58.4           
      (2, 16, 32)         |             58.7           
      (2, 16, 32, 32)     |             68.1           
      (4, 2, 16, 32, 32)  |             65.5           

Times are in microseconds (us).

[ masked_scatter_ mdtype torch.bool input dtypetorch.float64 ]
                          |  torch.masked_scatter_ CUDA
1 threads: --------------------------------------------
      (32,)               |             56.4           
      (32, 32)            |             62.0           
      (2, 16, 32)         |             58.9           
      (2, 16, 32, 32)     |             72.0           
      (4, 2, 16, 32, 32)  |             70.8           

Times are in microseconds (us).
```

**After PR**
```
Using pytorch 1.8.0a0+e44b2b7
[ masked_scatter_ mdtype torch.uint8 input dtypetorch.int32 ]
                          |  torch.masked_scatter_ CUDA
1 threads: --------------------------------------------
      (32,)               |             81.3           
      (32, 32)            |             80.6           
      (2, 16, 32)         |             81.2           
      (2, 16, 32, 32)     |             90.3           
      (4, 2, 16, 32, 32)  |             96.1           

Times are in microseconds (us).

[ masked_scatter_ mdtype torch.uint8 input dtypetorch.float32 ]
                          |  torch.masked_scatter_ CUDA
1 threads: --------------------------------------------
      (32,)               |             82.2           
      (32, 32)            |             82.2           
      (2, 16, 32)         |             83.3           
      (2, 16, 32, 32)     |             92.2           
      (4, 2, 16, 32, 32)  |             92.6           

Times are in microseconds (us).

[ masked_scatter_ mdtype torch.uint8 input dtypetorch.float64 ]
                          |  torch.masked_scatter_ CUDA
1 threads: --------------------------------------------
      (32,)               |             81.3           
      (32, 32)            |             87.6           
      (2, 16, 32)         |             82.6           
      (2, 16, 32, 32)     |             91.9           
      (4, 2, 16, 32, 32)  |             93.8           

Times are in microseconds (us).

[ masked_scatter_ mdtype torch.bool input dtypetorch.int32 ]
                          |  torch.masked_scatter_ CUDA
1 threads: --------------------------------------------
      (32,)               |             54.5           
      (32, 32)            |             56.4           
      (2, 16, 32)         |             57.6           
      (2, 16, 32, 32)     |             69.0           
      (4, 2, 16, 32, 32)  |             66.7           

Times are in microseconds (us).

[ masked_scatter_ mdtype torch.bool input dtypetorch.float32 ]
                          |  torch.masked_scatter_ CUDA
1 threads: --------------------------------------------
      (32,)               |             57.0           
      (32, 32)            |             56.3           
      (2, 16, 32)         |             59.1           
      (2, 16, 32, 32)     |             66.0           
      (4, 2, 16, 32, 32)  |             63.6           

Times are in microseconds (us).

[ masked_scatter_ mdtype torch.bool input dtypetorch.float64 ]
                          |  torch.masked_scatter_ CUDA
1 threads: --------------------------------------------
      (32,)               |             55.5           
      (32, 32)            |             56.7           
      (2, 16, 32)         |             60.5           
      (2, 16, 32, 32)     |             64.6           
      (4, 2, 16, 32, 32)  |             61.5           

Times are in microseconds (us).
```

<details>
    <summary>Code</summary>

```python
import torch
import itertools
import time
from torch.utils.benchmark import Timer
from torch.utils.benchmark import Compare
import sys

print('Using pytorch %s' % (torch.__version__))

shapes = [(32,), (32, 32), (2, 16, 32), (2, 16, 32, 32), (4, 2, 16, 32, 32)]
results = []
repeats = 10

for m_dtype in [torch.uint8, torch.bool]:
    for dtype in [torch.int32, torch.float, torch.double]:
        for mat1_shape in shapes:
            mat1 = torch.zeros(*mat1_shape, dtype=dtype, device='cuda')
            mask = (torch.randn(mat1.shape, device='cuda') > 0).to(m_dtype)
            source = torch.ones(mask.sum(), dtype=dtype, device='cuda')

            tasks = [("mat1.masked_scatter_(mask, source)", "torch.masked_scatter_ CUDA")]
            timers = [Timer(stmt=stmt, label=f"masked_scatter_ mdtype {m_dtype} input dtype{dtype}", sub_label=f"{(mat1_shape)}", description=label, globals=globals()) for stmt, label in tasks]

            for i, timer in enumerate(timers * repeats):
                results.append(
                    timer.blocked_autorange()
                )
                print(f"\r{i + 1} / {len(timers) * repeats}", end="")
                sys.stdout.flush()
comparison = Compare(results)
comparison.print()
```

</details>

File: aten/src/ATen/native/cuda/IndexKernel.cu [hunk 1]
```diff
@@ -252,6 +258,112 @@ Tensor& take_out_cuda(Tensor& out, const Tensor& self, const Tensor& index) {
...
+  Tensor b_mask;
+  std::tie(b_mask) = expand_inplace(self, mask, "masked_scatter_");
+
+  if (b_mask.dtype() == ScalarType::Byte) {
+    TORCH_WARN("masked_scatter_ received a mask with dtype torch.uint8, this behavior is now deprecated," \
+            "please use a mask with dtype torch.bool instead.");
+  }
+
+  // `mask` and `self` must have the same number of elements
+  TORCH_CHECK(self.numel() == b_mask.numel(), "Number of elements of self != Number of elements in mask");
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/IndexKernel.cu:355]: similar to cpu PR, this check does not seem to be necessary. Consider unifying common checks in a follow-up PR

**Review by ngimel — COMMENTED** [review]
Please add benchmarks for bigger sizes, current benchmarking seems to be mostly independent of size, which means that it's mostly benchmarking overheads.

**Comment by kshitij12345** [comment]
@ngimel Gentle Ping

File: aten/src/ATen/native/cuda/IndexKernel.cu [hunk 2]
```diff
@@ -252,6 +258,103 @@ Tensor& take_out_cuda(Tensor& out, const Tensor& self, const Tensor& index) {
...
+  auto maskPrefixSum = at::empty_like(mask, mask.options().dtype(kLong));
+
+  auto allocator = THCThrustAllocator(globalContext().lazyInitCUDA());
+
+  thrust::device_ptr<mask_t> maskData(mask_cont.data_ptr<mask_t>());
+  thrust::device_ptr<int64_t> maskPrefixSumData(
+      maskPrefixSum.data_ptr<int64_t>());
+
+  thrust::exclusive_scan(
+      thrust::cuda::par(allocator).on(c10::cuda::getCurrentCUDAStream()),
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/IndexKernel.cu:286]: can you please check if it works for numel > 2**31? There used to be some bugs for large cumsums.

File: test/test_torch.py [hunk 3]
```diff
@@ -4521,6 +4478,55 @@ def test_scatter_add_bool(self, device):
                                             [False, True, False, True, False],
                                             [True, False, True, False, True]], device=device))
 
+    @onlyOnCPUAndCUDA
+    def test_masked_scatter(self, device):
+        with warnings.catch_warnings(record=True) as w:
+            warnings.simplefilter("always")
+            for maskType in [torch.uint8, torch.bool]:
+                for dt in torch.testing.get_all_dtypes():
```
Comment (ngimel) [inline test/test_torch.py:4486]: please use @dtypes decorator instead of a loop

File: test/test_torch.py [hunk 4]
```diff
@@ -4521,6 +4478,55 @@ def test_scatter_add_bool(self, device):
...
+                    dest.masked_scatter_(mask, src)
+                    j = 0
+                    for i in range(num_dest):
+                        if mask[i]:
+                            dest2[i] = src[j]
+                            j += 1
+                    self.assertEqual(dest, dest2, atol=0, rtol=0)
+
+                    # make source bigger than number of 1s in mask
+                    src = torch.tensor([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=dt, device=device)
```
Comment (ngimel) [inline test/test_torch.py:4512]: your src was bigger in the original test case, what's the point of this one?

**Review by ngimel — COMMENTED** [review]
Minor comments.

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**